### ISSUE TITLE
Assignment Submission / Retrieval

### ISSUE THREAD

**Issue by nixpulvis** [nixpulvis/bottlenose#8]
I need to think this over again, things mostly work as I'd expect right now, but there has been talk of edge cases I need to think about.

**Comment by blerner** [comment]
Security/data integrity problem: if I log in as myself, not in staff mode and not impersonating anyone, and I go to http://107.170.114.15:3000/courses/1/assignments/1/submissions/5, I get to see Sally0 Student's homework submission for course1, assignment1.  If I go to .../submissions/6, I get to see Sally1 Student's homework for course1, assignment1.

If I go to http://107.170.114.15:3000/courses/3/assignments/1/submissions/5, I see, well, I'm not sure whether it's the same dummy data, or whether it's just ignoring the course, but I do see a link "Back to HW 0", and if I click it, I get an error because there isn't course3, assignment1.

http://107.170.114.15:3000/courses/3/assignments/100000/submissions/5, I *still* see Sally0 Student's homework submission for course1 assignment1...and the "Back to HW 0" link takes me to courses/3/assignments/1, which seems to me to indicate that the assignments number is just being lost somehow.

**Comment by nixpulvis** [comment]
If you look at the `submissions_controller.rb` you'll see that the app simply loads the `Submission` by id, and ignores the rest of the params. This is why in Rails in general deeply nested routes like this are an anti-pattern [1]. As far as the oddness of this, while I agree it's weird, doesn't seem like a huge problem to be because the only way to get this to happen is to manually edit the URL.

[1] - http://thepugautomatic.com/2013/05/unnesting-routes-in-rails/

**Comment by nixpulvis** [comment]
In terms of the first issue, I'm not sure we're on the same page for the staff role. There is no staff mode. You are a `site_admin` which means we can access all pages basically. To see how the pages will act to students, impersonate a student account. The logic for redirecting when a page shouldn't be visible is https://github.com/nixpulvis/bottlenose/blob/master/app/models/submission.rb#L184-L189

**Comment by blerner** [comment]
Yes, I just found where in the code that lives, and realized my mistake. I thought I had to go through the staff pages to get access to student pages, not that I had ambient authority by virtue of being a site admin. I'm less worried by this "problem" now.

**Comment by nixpulvis** [comment]
Ok good. I'm digging into add the roles for registrations to allow permissions for staff pages now.

**Comment by blerner** [comment]
As for students manually editing URLs: of course they're gonna try that. Who wouldn't resist the chance to find out hidden information, or try to change their grade somehow? I can see that it's an anti-pattern, but is it feasible to "cross-check" the course, assignment, and submission IDs against each other and against the current user, and error if *anything* is out of sync?

**Comment by nixpulvis** [comment]
Yep, it is possible. And definatly something worth doing at some point. But there shouldn't be a leak of data here, just some possible confusion. The permissions should all be based on  the data that is asked to be loaded, not the URL directly. For example consider the url `/courses/2/assignments/3/submissions/4` This will make a call to `Submission.find(4)` and check if `current_user` can view this submission. This is independent of the fact that submission 4 might not belong to assignment 3 or course 2.
